libtiff/libtiff_ctypes.py

```diff
@@ -786,59 +831,54 @@ def read_tiles(self, dtype=np.uint8):
-            for y in range(0, num_irows, num_trows):
-                for x in range(0, num_icols, num_tcols):
-                    r = self.ReadTile(tmp_tile.ctypes.data, x, y, z, 0)
+        # this number includes extra samples
+        samples_pp = self.GetField('SamplesPerPixel')
+        if samples_pp is None:  # default is 1
+            samples_pp = 1
+        planar_config = self.GetField('PlanarConfig')
+        if planar_config is None:  # default is contig
+            planar_config = PLANARCONFIG_CONTIG
+
+        def read_plane(plane, plane_index, tmp_tile):
+            for y in xrange(0, num_irows, num_trows):
+                for x in xrange(0, num_icols, num_tcols):
+                    r = self.ReadTile(tmp_tile.ctypes.data, x, y, 0, plane_index)
                     if not r:
                         raise ValueError(
                             "Could not read tile x:%d,y:%d,z:%d from file" % (
                                 x, y, z))
 
-                    if ((y + num_trows) > num_irows) or (
-                                (x + num_tcols) > num_icols):
-                        # We only need part of the tile because we are on
-                        # the edge
-                        if num_idepth == 1 and num_irows == 1:
-                            full_image[x:x + num_tcols] = \
-                                tmp_tile[0,:num_icols - x]
-                        elif num_idepth == 1:
-                            full_image[y:y + num_trows,
-                                       x:x + num_tcols] = \
-                                tmp_tile[:num_irows - y,
-                                         :num_icols - x]
-                        else:
-                            full_image[z,
-                                       y:y + num_trows,
-                                       x:x + num_tcols] = \
-                                tmp_tile[:num_irows - y,
-                                         :num_icols - x]
-                    else:
-                        if num_idepth == 1 and num_irows == 1:
-                            full_image[x:x + num_tcols] = tmp_tile[0, :]
-                        elif num_idepth == 1:
-                            full_image[y:y + num_trows,
-                                       x:x + num_tcols] = tmp_tile[:, :]
-                        else:
-                            full_image[z, y:y + num_trows,
-                                       x:x + num_tcols] = tmp_tile[:, :]
+                    # if the tile is on the edge, it is smaller
+                    tile_width = min(num_tcols, num_icols - x)
+                    tile_height = min(num_trows, num_irows - y)
+
+                    plane[y:y + tile_height, x:x + tile_width] = \
+                        tmp_tile[:tile_height, :tile_width]
+
+
+        if samples_pp == 1:
+            # if there's only one sample per pixel there is only one plane
+            full_image = np.zeros((num_irows, num_icols), dtype=dtype)
```

Comment:
Instead of np.zeros(), use np.empty() which is a little bit faster (as it doesn't write 0's).
Also, instead of using np.ascontiguousarray(), just use argument "order='C'" in np.empty().